# Scripts generated while completing my PhD thesis

**Josh Atkins -- University of Newcastle 2017** 

**Supervisors: `Prof Murray Cairns and Dr Chantel Fitzsimmons`** 

email:
`joshua.atkins  at uon.edu.au`


## list of script 

**Chapter 2 - Arrays scripts** 

```
-Neuromol_QC - Automated Array QC script 

-VCF_Chunk - Phasing and impuation script 

-Merge_Plink - Putting chunks back into whole chromosomes

-PRS_group_comparison - Script to do the stats 

-CNV_scoring - Script to score CNVs 

-Stats scripts - collection of scripts used to perform the stats and logistic regression classifier

```

**Chapter 3 - Exome-seq and RNA-seq scripts**

```
-Exome_pipeline - Automated pipeline used for both the exome sequencing and whole genome sequencing 

-RNA_seq_pipeline - RNA sequencing pipeline 
```

**Chapter 4 - WGS and LoF**
 ```
-Database_scripts - python scripts used to add pLi scores, phenotype group info and put read counts for each variant across the groups

-vcf_to_Lof - The bash scripts/commands used to filter INDEL and SNP that are LoF and annotate them to dnNSFP

```
